In [1]:
# Install necessary libraries
!pip install -qU cassandra-driver langchain langchain_community langchain-google-genai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 49.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 52.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 69.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.3/41.3 kB 2.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 411.6/411.6 kB 28.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.3/49.3 kB 3.4 MB/s eta 0:00:00


In [3]:
# Import libraries
from cassandra.cluster import Cluster
from cassandra.auth import PlainTextAuthProvider
import pandas as pd
import uuid
import os

# --- DataStax Astra DB Setup ---
# Replace with your Astra DB credentials
ASTRA_DB_SECURE_CONNECT_BUNDLE_PATH = "secure-connect-supermindhackthon.zip"  # Provide correct path
ASTRA_DB_APPLICATION_TOKEN = "AstraCS:MycfYsKryCoASlLrckFEemiE:2e7c9606a0d4388be81474506ab3ae16244e78311879cee93370c91cd0d5c8f7"
ASTRA_DB_KEYSPACE = "posts"  # Replace with your keyspace name

# Connect to Astra DB
cloud_config = {
    'secure_connect_bundle': ASTRA_DB_SECURE_CONNECT_BUNDLE_PATH
}
auth_provider = PlainTextAuthProvider("token", ASTRA_DB_APPLICATION_TOKEN)
cluster = Cluster(cloud=cloud_config, auth_provider=auth_provider, protocol_version=4)
session = cluster.connect()

# Set the keyspace
session.set_keyspace(ASTRA_DB_KEYSPACE)

In [4]:
# --- Create Table ---
# Create a table with clustering order by post_type
session.execute(
    """
    CREATE TABLE IF NOT EXISTS social_media_engagement (
        post_id UUID,
        post_type TEXT,
        likes INT,
        shares INT,
        comments INT,
        PRIMARY KEY (post_id, post_type)
    ) WITH CLUSTERING ORDER BY (post_type ASC);
    """
)

# --- Sample Data Insertion ---
sample_data = [
    {'post_id': uuid.uuid4(), 'post_type': 'carousel', 'likes': 100, 'shares': 20, 'comments': 10},
    {'post_id': uuid.uuid4(), 'post_type': 'reels', 'likes': 200, 'shares': 40, 'comments': 30},
    {'post_id': uuid.uuid4(), 'post_type': 'static_image', 'likes': 50, 'shares': 10, 'comments': 5},
    {'post_id': uuid.uuid4(), 'post_type': 'carousel', 'likes': 110, 'shares': 22, 'comments': 12},
    {'post_id': uuid.uuid4(), 'post_type': 'reels', 'likes': 210, 'shares': 42, 'comments': 32},
]

for row in sample_data:
    session.execute(
        """
        INSERT INTO social_media_engagement (post_id, post_type, likes, shares, comments)
        VALUES (%s, %s, %s, %s, %s)
        """,
        (row['post_id'], row['post_type'], row['likes'], row['shares'], row['comments'])
    )

In [5]:

# --- Analyze Engagement ---
def analyze_engagement(post_type):
    query = """
    SELECT AVG(likes) AS avg_likes, AVG(shares) AS avg_shares, AVG(comments) AS avg_comments
    FROM social_media_engagement
    WHERE post_type = %s ALLOW FILTERING
    """
    result = session.execute(query, [post_type])
    row = result.one()
    if row:
        return row.avg_likes, row.avg_shares, row.avg_comments
    return None, None, None

# Example usage
post_type = "carousel"  # Replace with user input
avg_likes, avg_shares, avg_comments = analyze_engagement(post_type)

if avg_likes and avg_shares and avg_comments:
    print(f"Average Likes: {avg_likes}")
    print(f"Average Shares: {avg_shares}")
    print(f"Average Comments: {avg_comments}")

Average Likes: 105
Average Shares: 21
Average Comments: 11


In [7]:
# --- LangChain Google GenAI Integration ---
from langchain_google_genai import GoogleGenerativeAI
from langchain import PromptTemplate
# Initialize Google GenAI
llm = GoogleGenerativeAI(model="gemini-pro", google_api_key="AIzaSyDMV1Lq3SAWeaxromOag8vlJtOqKnOHqa4")

# Define Prompt Template
template = """Analyze the following social media engagement data:
Average Likes: {avg_likes}
Average Shares: {avg_shares}
Average Comments: {avg_comments}

Consider this data to be representative of {post_type} posts. Compare it to hypothetical data for other formats, such as static images, videos, and text posts. Provide concise insights on the strengths and weaknesses of {post_type} posts compared to others. (example insights like Carousel posts have 20% higher engagement than static posts, Reels drive 2x more comments compared to other formats, give in markdown format with emojies)"""

prompt = PromptTemplate(
    input_variables=["avg_likes", "avg_shares", "avg_comments", "post_type"],
    template=template
)
print(prompt.format(avg_likes=avg_likes, avg_shares=avg_shares, avg_comments=avg_comments, post_type=post_type))

# Generate Insights
if avg_likes is None or avg_shares is None or avg_comments is None:
    print(f"Insufficient data to analyze post type: {post_type}")
else:
    insights = llm(prompt.format(
        avg_likes=avg_likes,
        avg_shares=avg_shares,
        avg_comments=avg_comments,
        post_type=post_type
    ))
    print(insights)


# --- Close the Connection ---
session.shutdown()
cluster.shutdown()


Analyze the following social media engagement data:
Average Likes: 105
Average Shares: 21
Average Comments: 11

Consider this data to be representative of carousel posts. Compare it to hypothetical data for other formats, such as static images, videos, and text posts. Provide concise insights on the strengths and weaknesses of carousel posts compared to others. (example insights like Carousel posts have 20% higher engagement than static posts, Reels drive 2x more comments compared to other formats, give in markdown format with emojies)
**Carousel Posts vs. Other Formats**

* **Likes:** Carousel posts have a relatively high average number of likes (105), indicating that they are generally well-received by audiences.
* **Shares:** Carousel posts have a moderate average number of shares (21), suggesting that they are not as widely shared as other formats.
* **Comments:** Carousel posts have a low average number of comments (11), indicating that they do not generate as much discussion as o

**Carousel Posts vs. Other Formats**

* **Likes:** Carousel posts have a relatively high average number of likes (105), indicating that they are generally well-received by audiences.
* **Shares:** Carousel posts have a moderate average number of shares (21), suggesting that they are not as widely shared as other formats.
* **Comments:** Carousel posts have a low average number of comments (11), indicating that they do not generate as much discussion as other formats.

**Comparison to Hypothetical Data for Other Formats:**

**Static Images:**
* Average Likes: 80
* Average Shares: 25
* Average Comments: 15

**Videos:**
* Average Likes: 120
* Average Shares: 30
* Average Comments: 20

**Text Posts:**
* Average Likes: 90
* Average Shares: 18
* Average Comments: 12

**Insights:**

* **Carousel posts:** Excel in likes but fall short in shares and comments compared to other formats. They provide a visually engaging experience but may not encourage as much discussion.
* **Static images:** Perform well in shares but have lower likes and comments than carousel posts. They are suitable for quick and concise messages.
* **Videos:** Drive the highest engagement in all three metrics. They are highly captivating and foster interaction.
* **Text posts:** Offer a balance of likes and comments but have lower shares. They are effective for sharing information and sparking conversations.